File: lib/netzke/widget/border_layout_panel.rb
module Netzke
  module Widget
    
    # Panel with border layout
    # == Example configuration:
    # 
    #     :items => [
    #       {:title => "Item One", :class_name => "Widget::Panel", :region => :center},
    #       {:title => "Item Two", :class_name => "Widget::Panel", :region => :west, :width => 300, :split => true}
    #     ]
    class BorderLayoutPanel < Base
      
      include Container
      
      def self.js_properties
        {
          :layout => 'border',
          :init_component => <<-END_OF_JAVASCRIPT.l,
            function(){
              Ext.each(['center', 'west', 'east', 'south', 'north'], function(r){
                // A function to access a region widget (even if the widget gets reloaded, the function will work).
                // E.g.: getEastWidget()
                this['get'+r.capitalize()+'Widget'] = function(){
                  return this.find('region', r)[0];
                }.createDelegate(this);
              }, this);

              // Now let Ext.Panel do the rest
              #{js_full_class_name}.superclass.initComponent.call(this);

              // First time on "afterlayout", set resize events
              if (this.persistentConfig) {this.on('afterlayout', this.setResizeEvents, this, {single: true});}
            }
          END_OF_JAVASCRIPT
          
        }
      end

    end
  end
end

File: lib/netzke/basepack/border_layout_panel/javascripts/border_layout_panel.js
{
  layout: 'border',

  initComponent: function(){
    this.callParent();

    // First time on "afterlayout", set resize events
    if (this.persistence) {this.on('afterlayout', this.setRegionEvents, this, {single: true});}
  },

  setRegionEvents: function(){
    this.items.each(function(item, index, length){
      if (!item.oldSize) item.oldSize = item.getSize(); // remember initial size

      item.on('resize', function(panel, w, h){
        if (panel.region !== 'center' && w && h) {
          var params = {region:panel.region};

          if (panel.oldSize.width != w) {
            params.width = w;
          } else {
            params.height = h;
          }

          panel.oldSize = panel.getSize();
          this.regionResized(params);
        }
      }, this);

      item.on('collapse', function(panel){
        this.regionCollapsed({region: panel.region});
      }, this);

      item.on('expand', function(panel){
        this.regionExpanded({region: panel.region});
      }, this);

    }, this);
  }
}
